Added MireDot as a REST API doc generator

idelpivnitskiy opened the pull request:
This work is based on [current discussion](http://aerogear-dev.1069024.n5.nabble.com/aerogear-dev-do-we-want-swagger-for-REST-endpoint-documentation-td11342.html).

I haven't removed jax-doclets and only added Miredot.

There are some examples of possible [documentation variants](http://condor.pp.ua/aerogear-ups-rest-api-doc/). In this source code I use the second variant (Multi page collapse all).

We are also able to:
* change title;
* hide `Issues` tab;
* add HTML snippet with introduction text.

Also I'm able to refactor current API realisation later. It will simplify and improve documentaion.

matzew commented on jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/metrics/DashboardEndpoint.java:
```diff
@@ -16,6 +16,7 @@
  */
 package org.jboss.aerogear.unifiedpush.rest.metrics;
 
+import com.qmino.miredot.annotations.ReturnType;
```
I thought there are no custom annottions :)

idelpivnitskiy replied:
:) it is impossible to analyze `javax.ws.rs.core.Response`
I'm going to refactor some API methods to return real objects instead of  `javax.ws.rs.core.Response` and use [ExceptionMapper](https://docs.oracle.com/javaee/6/api/javax/ws/rs/ext/ExceptionMapper.html) for other response codes, such 404. It will help us to minimize usage of `@ReturnType`.

matzew commented:
Relevant JIRA: https://issues.jboss.org/browse/AGPUSH-1366 

@idelpivnitskiy do you have a JIRA account already? Feel free to assign the above ticket to you, and link from there to this PR

nice work!!

idelpivnitskiy commented:
@matzew thanks!
Yes, my JIRA account is the same as GitHub account. I've added PR to this issue, but I think that I haven't got permissions to assign it to me.

idelpivnitskiy commented:
@matzew Which variant of  presented docs do you prefer? And what do you think about custom titile, HTML snippet, issues tab? See my first comment.

matzew commented:
> Yes, my JIRA account is the same as GitHub account. I've added PR to this issue, but I think that I haven't got permissions to assign it to me.

hrm, need to check what's going on

matzew commented:
@idelpivnitskiy can you rebase your branch, on latest changes on master ?

idelpivnitskiy commented:
@matzew done :)

idelpivnitskiy commented:
@matzew I hurried to merge your commit. If you want to fully remove jaxrs doclet, we also have to remove doclet's specific javadoc tags, like `@HTTP`. I suggest to remove your current commit from my branch (with force push), update your commit and open a new PR to my branch.

matzew commented on jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/sender/PushNotificationSenderEndpoint.java:
```diff
@@ -100,9 +101,17 @@
     @POST
     @Consumes(MediaType.APPLICATION_JSON)
+    @ReturnType("java.lang.String")
     public Response send(final UnifiedPushMessage message, @Context HttpServletRequest request) {
```
this generates more doc than we want, in this case.

For instance the UnifiedPushMessage has `ipAddress` for internal tracking - that is not exposed, see:
https://aerogear.org/docs/specs/aerogear-unifiedpush-rest/sender/index.html

idelpivnitskiy replied:
Hm... I think that in this case we should split `UnifiedPushMessage` to `UnifiedPushMessage` and something like `UnifiedPushMessageExtended extends UnifiedPushMessage`:
```java
@BodyType("org.jboss.aerogear.unifiedpush.message.UnifiedPushMessage")
public Response send(final UnifiedPushMessageExtended message, @Context HttpServletRequest request) {
...
}
```

matzew replied:
> Hm... I think that in this case we should split UnifiedPushMessage to UnifiedPushMessage and something like UnifiedPushMessageExtended extends UnifiedPushMessage

@idelpivnitskiy how about having an inherited class inside of the push/rest module -> `InternalUnifiedPushMEssage extends UnifiedPushMessage`

@sebastienblanc I think that extra class would be also handy on the analtycs tracking stuff, right?

sebastienblanc replied:
+1, Is tracked here https://issues.jboss.org/browse/AGPUSH-1381

idelpivnitskiy commented:
@matzew @sebastienblanc I've split UnifiedPushMessage and for the quick fix I've removed `ipAddress` from test json files. Is it ok? Should I override `UnifiedPushMessage.toStrippedJsonString()`?

sebastienblanc commented:
I would prefer to see that in a separate PR.

Envoyé de mon iPhone

> Le 4 mai 2015 à 02:32, Idel Pivnitskiy <notifications@github.com> a écrit :
> 
> @matzew @sebastienblanc I've split UnifiedPushMessage and for the quick fix I've removed ipAddress from test json files. Is it ok? Should I override UnifiedPushMessage.toStrippedJsonString()?
> 
> —
> Reply to this email directly or view it on GitHub.
>

idelpivnitskiy commented:
@sebastienblanc done
I'll update miredot documentation for the `send` method after merging #542

idelpivnitskiy commented:
@matzew @sebastienblanc updated http://condor.pp.ua/aerogear-ups-rest-api-doc/latest/index.html

matzew commented:
looks very nice !

once we have the analytics thing in, I will address some of your comments
on this PR, sounds good ?

On Thu, May 7, 2015 at 12:03 PM, Idel Pivnitskiy <notifications@github.com>
wrote:

> @matzew <https://github.com/matzew> @sebastienblanc
> <https://github.com/sebastienblanc> updated
> http://condor.pp.ua/aerogear-ups-rest-api-doc/latest/index.html
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/526#issuecomment-99801876>
> .
>



-- 
Matthias Wessendorf

blog: http://matthiaswessendorf.wordpress.com/
sessions: http://www.slideshare.net/mwessendorf
twitter: http://twitter.com/mwessendorf

idelpivnitskiy commented:
@matzew ok

matzew commented:
Closing in fav. of https://github.com/aerogear/aerogear-unifiedpush-server/pull/550
